### PR TITLE
OHRM5X-1260 : Migrations 4.0 to 4.3.3

### PR DESCRIPTION
**Pull request by SharukaNP**
At the moment only upgrades from 4.3 upwards are reliable.

### CODE REVIEW

File: installer/Migration/V4_3_3/Migration.php [hunk 1]
```diff
@@ -29,30 +31,118 @@ class Migration extends AbstractMigration
      */
     public function up(): void
     {
-        $this->getSchemaHelper()->changeColumn('hs_hr_employee', 'emp_number', [
-            'Type' => Types::INTEGER, 'Length' => 7, 'Notnull' => true, 'Autoincrement' => true
-        ]);
+        /*$this->getSchemaHelper()->changeColumn('hs_hr_employee', 'emp_number', [
```
Comment (SharukaNP) [inline installer/Migration/V4_3_3/Migration.php:34]: @RajithaKumara  here we have a small issue adding auto Increment to an existing column
Shall I make a function to update those existing data?

File: installer/Migration/V4_1_1/Migration.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     * @inheritDoc
+     */
+    public function up(): void
+    {
+        $this->getSchemaHelper()->dropForeignKeys('hs_hr_emp_member_detail', ['hs_hr_emp_member_detail_ibfk_1']);
+        $this->getSchemaHelper()->dropForeignKeys('hs_hr_emp_member_detail', ['hs_hr_emp_member_detail_ibfk_2']);
+
+        $this->getSchemaHelper()->dropPrimaryKey('hs_hr_emp_member_detail');
+
+        $this->getSchemaHelper()->addColumn('hs_hr_emp_member_detail', 'id', Types::INTEGER, ['Length' => '6', 'Notnull' => false]);
```
Comment (SharukaNP) [inline installer/Migration/V4_1_1/Migration.php:38]: Here also the autoincrement issue

**Comment by SharukaNP** [comment]
DB comparability for setting FK constraints on and off

File: installer/Migration/V4_3_3/Migration.php (hunk 1 above)
Comment (SharukaNP) [reply installer/Migration/V4_3_3/Migration.php]: Solved by using raw sql

File: installer/Migration/V4_1_1/Migration.php (hunk 2 above)
Comment (SharukaNP) [reply installer/Migration/V4_1_1/Migration.php]: Solved by using raw sql

**Comment by SharukaNP** [comment]
From version 4.00 - 4.1.1 there are issues when using just he migrations due to the instance.Identifier.

File: installer/Migration/V4_7_0/Migration.php [hunk 3]
```diff
@@ -28,123 +28,7 @@ class Migration extends AbstractMigration
...
-            'translateUniqueId',
-            ['lang_string_id', 'language_id']
-        );
-        $this->getSchemaHelper()->getSchemaManager()->createUniqueConstraint($uniqueConstraint, 'ohrm_i18n_translate');
-        $this->getDataGroupHelper()->insertScreenPermissions(__DIR__ . '/permission/screen.yaml');
-        $languages = $this->readlanguageYaml(__DIR__ . '/language/languages.yaml');
-        $this->insertLanguages($languages);
-        $groups = $this->readlanguageYaml(__DIR__ . '/language/groups.yaml');
-        $this->insertGroups($groups);
+        // no db chnages for this version
```
Comment (RajithaKumara) [inline installer/Migration/V4_7_0/Migration.php:31]: double check all the migrations files, to confirm whether this kind of mistakes exists

File: installer/Util/V1/SchemaHelper.php [hunk 4]
```diff
@@ -41,11 +45,22 @@ public function __construct(AbstractSchemaManager $schemaManager)
...
     /**
-     * @return AbstractSchemaManager
+     * @return Connection|null
      */
-    public function getSchemaManager(): AbstractSchemaManager
+    private function getConnection(): ?Connection
     {
-        return $this->schemaManager;
+        return $this->getEntityManager()->getConnection()->getWrappedConnection();
+    }
```
Comment (RajithaKumara) [inline installer/Util/V1/SchemaHelper.php:53]: this is wrong. Don't use `EntityManagerHelperTrait` in installer. need to discuss this

File: installer/Util/V1/SchemaHelper.php [hunk 5]
```diff
@@ -169,4 +181,47 @@ public function tableExists(array $table): bool
...
+     */
+    private function _getDbConnection(): Connection
+    {
+        return $this->getEntityManager()->getConnection()->getWrappedConnection();
+    }
+
+    /**
+     * @return void
+     */
+    public function _disableConstraints(): void
```
Comment (RajithaKumara) [inline installer/Util/V1/SchemaHelper.php:206]:
```suggestion
    public function disableConstraints(): void
```

File: installer/Util/V1/SchemaHelper.php [hunk 6]
```diff
@@ -169,4 +181,47 @@ public function tableExists(array $table): bool
...
+    {
+        $pdo = $this->_getDbConnection();
+        $pdo->exec('SET FOREIGN_KEY_CHECKS=0;');
+    }
+
+    /**
+     * @return void
+     */
+    public function _enableConstraints(): void
```
Comment (RajithaKumara) [inline installer/Util/V1/SchemaHelper.php:215]:
```suggestion
    public function enableConstraints(): void
```

File: installer/Util/V1/SchemaHelper.php [hunk 7]
```diff
@@ -169,4 +181,47 @@ public function tableExists(array $table): bool
...
+    {
+        $pdo = $this->_getDbConnection();
+        $pdo->exec('SET FOREIGN_KEY_CHECKS=1;');
+    }
+
+    public function execSql(array $sql)
```
Comment (RajithaKumara) [inline installer/Util/V1/SchemaHelper.php:221]:
```suggestion
    public function execSql(array $queries): void
```
add doc comment

File: installer/Util/V1/SchemaHelper.php [hunk 8]
```diff
@@ -169,4 +181,47 @@ public function tableExists(array $table): bool
...
+    {
+        foreach ((array)$sql as $query) {
```
Comment (RajithaKumara) [inline installer/Util/V1/SchemaHelper.php:223]:
```suggestion
        foreach ($sql as $query) {
```

File: installer/Util/V1/SchemaHelper.php [hunk 9]
```diff
@@ -169,4 +181,47 @@ public function tableExists(array $table): bool
...
+    public function dropPrimaryKey(string $tableName): void
+    {
+        $table = $this->getSchemaManager()->listTableDetails($tableName);
+        $this->getSchemaManager()->dropIndex($table->getPrimaryKey(), $table);
+    }
+
+    /**
+     * @return Connection
+     */
+    private function _getDbConnection(): Connection
```
Comment (RajithaKumara) [inline installer/Util/V1/SchemaHelper.php:198]: return type also need to discuss

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: installer/Migration/V4_7_0/Migration.php (hunk 3 above)
Comment (SharukaNP) [reply installer/Migration/V4_7_0/Migration.php:31]: Checked

**Comment by SharukaNP** [comment]
Added Migration Mapping. Should be good to merge.

File: installer/Migration/V4_3/permission/screen.yaml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+'Marcket Place Home Page':
```
Comment (RajithaKumara) [inline installer/Migration/V4_3/permission/screen.yaml:1]: Typo, to keep backward compatibility with 4.x

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**